internal/plugin/config_plugin.go

```diff
@@ -120,16 +122,61 @@ func (c *Config) parseInstanceConfiguration(correlationID string, instance *inst
 }
 
 func (c *Config) updateInstanceConfig(request *model.InstanceConfigUpdateRequest) {
+	// remove when tenantID is being set
+	tenantID := "7332d596-d2e6-4d1e-9e75-70f91ef9bd0e"
+
 	slog.Debug("Updating instance configuration")
 	instanceID := request.Instance.GetInstanceId()
 	if c.configServices[instanceID] == nil {
 		c.configServices[instanceID] = service.NewConfigService(request.Instance, c.agentConfig)
 	}
 
-	status := c.configServices[request.Instance.GetInstanceId()].UpdateInstanceConfiguration(
+	inProgressStatus := &instances.ConfigurationStatus{
+		InstanceId:    instanceID,
+		CorrelationId: request.CorrelationID,
+		Status:        instances.Status_IN_PROGRESS,
+		Timestamp:     timestamppb.Now(),
+		Message:       "Instance configuration update in progress",
+	}
+	c.messagePipe.Process(&bus.Message{Topic: bus.InstanceConfigUpdateTopic, Data: inProgressStatus})
+
+	skippedFiles, status := c.configServices[request.Instance.GetInstanceId()].UpdateInstanceConfiguration(
 		c.messagePipe.Context(),
 		request.CorrelationID,
 		request.Location,
 	)
-	c.messagePipe.Process(&bus.Message{Topic: bus.InstanceConfigUpdateCompleteTopic, Data: status})
+	c.messagePipe.Process(&bus.Message{Topic: bus.InstanceConfigUpdateTopic, Data: status})
+
+	if status.GetStatus() == instances.Status_FAILED {
+		rollbackInProgress := &instances.ConfigurationStatus{
+			InstanceId:    instanceID,
+			CorrelationId: request.CorrelationID,
+			Status:        instances.Status_ROLLBACK_IN_PROGRESS,
+			Timestamp:     timestamppb.Now(),
+			Message:       "Rollback in progress",
+		}
+		c.messagePipe.Process(&bus.Message{Topic: bus.InstanceConfigUpdateTopic, Data: rollbackInProgress})
+
+		err := c.configServices[instanceID].Rollback(c.messagePipe.Context(), skippedFiles,
```

Comment:
Adding a general comment here about our practice for passing contexts, feel free to close this thread before merging.

So I think we may potentially need to refactor our usage of `context.Context` inside our plugins. What we currently do is that we save the `ctx` to the `messagePipe` struct and we pass that same context to various functions that require a `context.Context`. Contexts should be request-bounded and should not be saved to structs: https://pkg.go.dev/context

A few solutions that come to mind: either always create a context with `context.WithCancel()` and append that cancel function to the `messagePipe` OR we keep the context in `messagePipe` and use that as a parent context for each request.

Thoughts @dhurley, @oliveromahony, @aphralG, @Dean-Coakley?

Reply:
I would have thought that on the intialization of the plugins we would create a new context from that, where the message pipe context is the parent. But based on that link this appears to be the incorrect way to do it. I guess we should widen the plugin interface to accept contexts then?

Reply:
> I guess we should widen the plugin interface to accept contexts then?

I'd say so if we want to conform to Go's conventions. It'll be important to pass contexts around correctly, as stuff like the correlation ID and tenant ID will likely be passed via the context.

Reply:
Do I need to make the changes in this PR or will there be a follow up PR ?

Reply:
Let's do a follow-up PR, this is outside the scope of what you were doing.